Automattic/wp-calypso
# Enhance Ad Tracking (fix: currency conversion, signup race condition)

### Changes proposed in this Pull Request

#### Fixes

- Fix bug:`'unknown ' === cookies.country_code` (note the space).
- Fix `analytics.ga.recordEvent()` by converting the optional `value` param to an integer before firing the event. GA refuses to accept events that try to pass a numeric value with decimals. A value should be an integer. This correction should fix a discrepancy; i.e., the missing `calypso_checkout_payment_success` events in GA.
- Fix cookie domain (use `.[root hostname]`) for SEM `ad_details` cookie.

### Enhancements

- Add `recordSocialRegistration()`, which will handle Google (aka: Social) user registration tracking. Also, we'll start tracking this event in Google Analytics now too, in addition to Tracks.
- On `calypso_signup_complete`, use a delay of 1000-3000 ms when tracking signups. There are several ad trackers and we need to give them time to load before redirecting.
- On `calypso_signup_complete`, the `is_new_user` prop is now _also_ `true` when an _existing user_ meets the following conditions: (a) registered <= 7 days ago, (b) currently has no sites, (c) is creating a new site now. This way we pick up new-ish users who return to create a site after having originally been through an account-only flow.

### Exchange Rates

- Add an up-to-date list of `EXCHANGE_RATES` for ad tracking revenue calculations. This is a short-term fix, with a long-term fix being one that would collect these from an up-to-date source.

### Testing instructions

In an incognito window, open [Calypso live URL w/ `?flags=gdpr-banner,google-analytics,ad-tracking`](https://hash-88b3d1944f0affef4ef5df4b4e777efef6754dff.calypso.live/start/user?flags=gdpr-banner,google-analytics,ad-tracking)

Set Console tab and Network tab to "Preserve log".

In Console tab, enter each of these lines and reload the page.
```
localStorage.setItem( 'debug', 'calypso:*' ); // Show everything.
document.cookie = 'flags=gdpr-banner,google-analytics,ad-tracking';
document.cookie = 'sensitive_pixel_option=yes';
document.cookie = 'country_code=DE';
```

---

- In Console tab, confirm `isAdTrackingAllowed: true`
- Confirm cookie banner is NOT displayed when testing from `country_code=DE`.

---

On initial pageview, confirm:

- With no filters applied in Console, confirm no JS errors.
- Filter Console tab by `calypso:analytics:ad-tracking`, and confirm tracking scripts are loading and that pageview and Signup Start trackers are all firing.
  - It's a good idea to confirm each of the trackers in the Network tab also.

![Screen Shot 2019-04-11 at 09 35 06](https://user-images.githubusercontent.com/1563559/55979856-97bdbc80-5c3f-11e9-9f71-945c6e6221c3.png)


---

Complete signup, choosing the "Free" option and confirm:

- No JS errors in console.

- Filter console by `calypso:analytics:ad-tracking` and confirm Registration trackers are firing.
  - It's a good idea to confirm each of the trackers in the Network tab also.

![Screen Shot 2019-04-11 at 09 42 47](https://user-images.githubusercontent.com/1563559/55979900-b91ea880-5c3f-11e9-85a0-a502fdfb1827.png)

- Filter console by regex `/calypso:signup|ad-tracking/` and confirm Signup trackers are firing inside of these two debug markers, which are new in this PR.
  - It's a good idea to confirm each of the trackers in the Network tab also.

![Annotation on 2019-3-4](https://user-images.githubusercontent.com/1563559/55979932-ccca0f00-5c3f-11e9-849d-3667f2531013.png)

![Screen Shot 2019-04-11 at 09 52 19](https://user-images.githubusercontent.com/1563559/55979866-9e4c3400-5c3f-11e9-843f-6970915682d8.png)

---

Repeat the same test, but this time modify the approach.

- Begin from `/start/account` (account-only flow).
- After creating an account, reload the page, return to create a site and confirm the `calypso_signup_complete` fires with `is_new_user: true`, `is_new_site: true`.

---

Repeat the same test, but this time modify the approach.

- Begin from `/start/account` (account-only flow).
- Create a test Google account and start your account using Google (social).
- Confirm that `calypso_user_registation_social_complete` is firing in Tracks & GA.
- After creating an account, reload the page, return to create a site and confirm the `calypso_signup_complete` fires with `is_new_user: true`, `is_new_site: true`.

---

Repeat with each type of paid plan to confirm no problems. When you arrive at checkout, add `?flags=gdpr-banner,google-analytics,ad-tracking` to end of the URL, refresh, and confirm `isAdTrackingAllowed: true`.

Use a test card (`4242 4242 4242 4242`) or Points and confirm:

- No JS errors in console.
- All of the above are still true.

_See screenshots above._

Additionally, confirm that:

- Filter console by `calypso:analytics:ad-tracking` and confirm Purchase trackers are firing.
  - It's a good idea to confirm each of the trackers in the Network tab also.
- Confirm that all USD amounts are as expected in each of the trackers; i.e., that currency conversion is continuing to work as we expect it to.

![Screen Shot 2019-04-11 at 10 02 37](https://user-images.githubusercontent.com/1563559/55980476-02bbc300-5c41-11e9-8354-6ca3ba7aa3da.png)

Filter console by `calypso_checkout_payment_success` and confirm that an integer value is sent in the Google Analytics event. Also confirm in the Network tab.

![Screen Shot 2019-04-11 at 10 08 10](https://user-images.githubusercontent.com/1563559/55980854-d785a380-5c41-11e9-9a47-7c64d98aa381.png)

---

Repeat purchase test with different currencies and let me know how you did it? ha
